Sources/AudioKitUI/Visualizations/SpectrogramFlatView/SpectrogramModel.swift

```diff
@@ -0,0 +1,173 @@
+// Copyright AudioKit. All Rights Reserved. Revision History at http://github.com/AudioKit/AudioKitUI/
+//
+
+import AudioKit
+import SwiftUI
+
+struct SpectrogramFFTMetaData {
+    // fftSize defines how detailled the music is analyzed in the time domain. 
+    // the lower the value, the less detail:
+    // * 1024: will receive about four analyzed frequencies between C2 and C* (65Hz to 130Hz). 
+    //             New data comes roughly 21.5 times per second, each 46ms.  
+    // * 2048: will receive about eight analyzed frequencies between C2 and C* (65Hz to 130Hz).  
+    //            New data comes roughly 11 times per second, each 93ms. 
+    // * 4096: will receive about 16 analyzed frequencies between C2 and C* (65Hz to 130Hz). 
+    //            New data comes roughly  5.5 times per second, each 186ms.  
+    // Choose a higher value when you want to analyze low frequencies, 
+    // choose a lower value when you want fast response and high frame rate on display.
+    // Default: 2048
+    var fftSize = 4096/2
+
+    // how/why can the sample rate be edited? Shouldn't this come from the node/engine?
+    // if the sample rate is changed, does the displayed frequency range also have to be changed?
+    // took this from existing SpectrogramView, will investigate later
+    var sampleRate: double_t = 44100
+}
+
+@available(iOS 17.0, *)
+struct SliceQueue {
+    var maxItems: Int = 120
+    var items: [SpectrogramSlice] = []
+
+    public mutating func pushToQueue(element: SpectrogramSlice) {
+        enqueue(element: element)
+        if items.count > maxItems {
+            dequeue()
+        }
+    }
+
+    private mutating func enqueue(element: SpectrogramSlice) {
+        items.append(element)
+    }
+
+    private mutating func dequeue() {
+        if !items.isEmpty {
+            items.remove(at: 0)
+        }
+    }
+}
+
+///  Model for the SpectrogramFlatView. Makes connection to the audio node and receives FFT data
+@available(iOS 17.0, *)
+class SpectrogramFlatModel: ObservableObject {
+    @Published var slices = SliceQueue()
+    var sliceSize = CGSize(width: 10, height: 250) {
+        didSet {
+            if xcodePreview { createTestData() }
+        }
+    }
+    let nodeMetaData = SpectrogramFFTMetaData()
+    let xcodePreview = ProcessInfo.processInfo.environment["XCODE_RUNNING_FOR_PREVIEWS"] == "1"
+    var nodeTap: FFTTap!
+    var node: Node?
+
+    // we use 48Hz, which is a bit lower than G1. A1 would be 440Hz/8 = 55Hz.  
+    // The lowest human bass voice in choral music is reaching down to C1 (32.7 Hz). 
+    // don't go lower than 6.0, it just doesn't make sense and the display gets terribly distorted
+    // don't use 0 as it breaks the display because log10(0) is undefined and this error not handled
+    var minFreq: CGFloat = 48.0
```

Comment:
do these values change? otherwise move to lets.

Reply:
Yes, totally. Moved it not only to let but also to SpectrogramFFTMetaData so it can be opened up as public parameter in future. Ideally the minFreq and maxFreq can be configured on runtime. [3868d03](https://github.com/AudioKit/AudioKitUI/pull/82/commits/3868d039b678f2b4816e438f2a68ec924cc4da31)